drivers/usb/usbstor_new/disk.c

```diff
@@ -55,42 +121,29 @@ USBSTOR_HandleInternalDeviceControl(
         {
             DPRINT("SRB_FUNCTION_EXECUTE_SCSI\n");
 
-            //
-            // check if request is valid
-            //
-            if (Request->SrbFlags & (SRB_FLAGS_DATA_IN | SRB_FLAGS_DATA_OUT))
+            if (!IsRequestValid(Irp))
             {
-                //
-                // data is transferred with this irp
-                //
-                if ((Request->SrbFlags & (SRB_FLAGS_DATA_IN | SRB_FLAGS_DATA_OUT)) == (SRB_FLAGS_DATA_IN | SRB_FLAGS_DATA_OUT) ||
-                    Request->DataTransferLength == 0 ||
-                    Irp->MdlAddress == NULL)
-                {
-                    //
-                    // invalid parameter
-                    //
-                    Status = STATUS_INVALID_PARAMETER;
-                    break;
-                }
+                DPRINT1("USBSTOR_HandleInternalDeviceControl: Bad Srb!\n");
+                Status = STATUS_INVALID_PARAMETER;
+                Irp->IoStatus.Status = Status;
+                IoCompleteRequest(Irp, IO_NO_INCREMENT);
+                return Status;
             }
-            else
+
+            if (Request->Cdb[0] == SCSIOP_MODE_SENSE)
             {
-                //
-                // sense buffer request
-                //
-                if (Request->DataTransferLength ||
-                    Request->DataBuffer ||
-                    Irp->MdlAddress)
-                {
-                    //
-                    // invalid parameter
-                    //
-                    Status = STATUS_INVALID_PARAMETER;
-                    break;
-                }
+                DPRINT("USBSTOR_Scsi: SRB_FUNCTION_EXECUTE_SCSI - FIXME SCSIOP_MODE_SENSE\n");
+                // FIXME Get from registry WriteProtect for StorageDevicePolicies;
+                // L"\\Registry\\Machine\\System\\CurrentControlSet\\Control\\StorageDevicePolicies"
+                // QueryTable[0].Name = L"WriteProtect"
             }
 
+            IoStack->Parameters.Others.Argument2 = PDODeviceExtension;
```

Comment:
Does this need to be in the stack location? Normally Irp->DriverContext is used for this kind of thing.